active/0024-opentelemetry-trace-integration.md

```diff
@@ -0,0 +1,214 @@
+
+- [OpenTelemetry Erlang lib documentation](https://opentelemetry.io/docs/instrumentation/erlang/)
+- [OpenTelemetry main trace components overview](https://opentelemetry.io/docs/concepts/signals/traces)
+- [MQTT trace context specification](https://w3c.github.io/trace-context-mqtt/)
+- [Trace context header specification](https://www.w3.org/TR/trace-context-1/#tracestate-header)
+- [OpenTelemetry messaging system semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md)
+- [Draft PR](https://github.com/emqx/emqx/pull/11696)
+
+## Motivation
+
+OpenTelemetry distributed trace integration is the part of EMQX Product Road-map.
+
+## Design
+
+### Core concepts and tracing scope
+
+The core traceable entity for EMQX is a message. It means that one trace should be associated with one message.
+
+For example, a single HTTP request is a common traceable entity for a HTTP server. HTTP server instrumented with OpenTelemetry receives a HTTP request, extracts trace context from headers (e.g., `traceparent`, `tracestate` headers) and traces any processing steps (spans) up to sending a HTTP response back to the client, associating all the spans with the same trace ID (1 request - 1 trace ID).
+HTTP client after receiving the response may proceed executing some subsequent operations, tracing and linking them to the same trace ID.
+
+Somewhat analogously, EMQX instrumented with OpenTelemetry, is expected to receive a published message, extract trace context (e.g., `traceparent`, `tracestate` User-Properties), and trace some/all processing steps under the same trace ID.
+Producer/consumer of the message may proceed tracing any subsequent operations relating them to the same trace ID.
+
+These traced steps (or spans) should include the following (**TBD**):
+
+- Publish message (traced by a node that received a published message)
+
+  - Send message to data bridges (traced by a node that received a published message)
+  
+    NOTE: if the external data system and/or communication protocol supports OpenTelemetry distributed tracing, trace context must be also propagated to it. EMQX bridge app should implement a function to inject the trace context into its own message. For example, the web-hook bridge should add `traceparent` / `tracestate` HTTP headers, so that the receiving end can continue tracing under the same trace ID.
+
+  - Route message (traced by a node that received a published message)
+
+    - Dispatch message (traced by all nodes that have matched subscribers)
+
+      - Deliver/send message to a subscriber (traced by all nodes that have matched subscribers)
+
+NOTE: the above list may be extended/changed, for now it follows the product requirements.
+
+Any other processing steps/events like client connection, authentication, subscription are currently not considered for OpenTelemetry tracing (**TBD**) due to the following reasons:
```

Comment:
Have we decided on:
* tracing authorization (this seems to be a part of publish flow)
* tracing puback, pubrel, pubcomp within the same trace id as publishes
* the scope/kind of events we are going to add to the traces?

Reply:
Not yet, but the latest result of the discussion with @zmstone is to start with a minimum number of trace spans, e.g.: 
- process a published message (from the point when the packet is received and decoded till it is dispatched locally and forwarded to other nodes)
- publish the message to a subscriber (one span per each connection on every node, starting from receiving the message by a connection process till sending it to a port).